### PR TITLE
Permissions module - first iteration

### PR DESCRIPTION
**Pull request by dmitrizagidulin**
This PR is for a long-running feature branch, `permissions`. ACL-related code (to read and write `.acl` resources from the client) will go into here first. Once we have consensus on the first iteration (agreement with the spec, feedback from devs, etc), it'll make its way into `master`.

### CODE REVIEW

**Comment by dmitrizagidulin** [comment]
(rebased this branch on latest master)

**Comment by dmitrizagidulin** [comment]
@ghanemabdo - don't forget to rebase any further code on this branch.
Specifically, assuming you have an `upstream` remote that you set up for this repo, via

`git remote add upstream https://github.com/solid/solid.js.git`

then you should do:

```
git fetch upstream
git rebase upstream/permissions
```

Ping me if you run into any conflicts or anything.

**Comment by dmitrizagidulin** [comment]
(Since it's relevant to this feature) Here's a PR to a solid-specific Web Access Control spec: https://github.com/solid/web-access-control-spec/pull/1

**Comment by deiu** [comment]
Friendly reminder that this PR is still failing to pass tests.

**Comment by dmitrizagidulin** [comment]
Force-rebased `permissions` on latest `master`. @ghanemabdo - be sure to check out a fresh copy of `permissions`, and work off of that.

File: lib/permissions.js [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+   * url of on ancestor of the resource.
+   * @method getAclResource
+   * @param url {String} The url of the resource to retrieve the acl resource
+   * url for
+   * @return {Promise<String>} The url of the acl resource
+   */
+  getAclResource: function getAclResource (url) {
+    var self = this
+
+    return new Promise(function (resolve, reject) {
```
Comment (dmitrizagidulin) [inline lib/permissions.js]: Functions that return Promises (such as `web.head(url)`) do not need to be wrapped in an additional Promise. The usage here should be:

```js
  getAclResource: function getAclResource (url) {
    return webClient.head(url)
      .then(function(response) {
        return response.acl
      })
  }
```

File: lib/permissions.js [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+      webClient.head(url).then(
+        function (solidResponse) {
+          var aclUrl
+          // if the resource exists and there is a logged in user, then
+          // succeed with the acl url
+          if (solidResponse.exists() && solidResponse.isLoggedIn()) {
```
Comment (dmitrizagidulin) [inline lib/permissions.js]: No need to check if the user is logged in - a `web.head()` request returns an acl link regardless of logged-in status.

**Comment by dmitrizagidulin** [comment]
@ghanemabdo - I added several sections to the WAC spec yesterday that are relevant to this PR, especially the [ACL Inheritance Algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) section.

**Comment by dmitrizagidulin** [comment]
(Rebased on master)

**Comment by dmitrizagidulin** [comment]
@ghanemabdo - I'm taking a look at the Permissions code currently, with an eye towards eventually merging this PR into master. Feedback as I go along:

* Let's rename `Owner` permission type to `Control`, since we don't have a notion of Owners in WAC.

**Comment by ghanemabdo** [comment]
@dmitrizagidulin I added a lot of modifications to this merge. Briefly:
* Added support to read and write the predicate `defaultForNew`. This is also reflected in the demo page.
* Added `Agent` object to encapsulate the information of an agent WebID, permissions, is a list or an individual, this authorization is `defaultForNew`
* Several bug fixes.

Remaining items:
1) Conforming to the new inheritance way in `getPermissions` and `getAclResource` where `GET` on an individual acl resource url always returns result.
2) Modifying the README file to reflect the most recent changes in permissions.js
3) Fix the bug when `accessTo` predicate has absolute url which doesn't work in some cases. Needs deeper investigation.

**Comment by dmitrizagidulin** [comment]
Just pushed some refactoring of the new code from today's PR. 
- Wrapped some more lines to 80 chars
- Removed unneeded `.catch()` statements
- Removed redundant `return new Promise(...)` usage (functions that return promises can be returned directly).

**Comment by dmitrizagidulin** [comment]
Rebased on current master/HEAD, squashed commits into one for easier rebasing in the future.

**Comment by dmitrizagidulin** [comment]
@ghanemabdo I have some thoughts for refactoring this permissions api. Let's 1) make the permissions list keyed by agent (instead of keyed by permission mode), and 2) make the API cascading. For example.

Proposed API:

```js
// First, we have our current getPermission() method.
solid.getPermissions('https://www.example.com/file1')  // <- already implemented
  .then(function (permissionSet) {
  })
// ^ is equivalent to:
solid.web.head('https://www.example.com/file1')  // or .get(...)
  .then(function (resource) {  // resource is an instance of the usual SolidResponse
    return resource.permissions()  // this figures out the .acl resource and fetches it, or throws a 401
  })
  .then(function (permissionSet) {
  })

// Ok, so, once we have our set of permissions, we can edit them via a chaining API:
var acl = solid.permission.acls
solid.getPermissions('https://www.example.com/file1')
  .then(function (permissionSet) {
    return permissionSet
      .addPermission(webId, acl.READ)
      .addPermission(webId2, [acl.READ, acl.WRITE])
      .addPermission(acl.EVERYONE, acl.READ)  // public / agentClass foaf:Agent
      .addGroupPermission(groupWebId, acl.READ)  // adds an agentClass entry
      .removePermission(webid3, acl.READ)
      .save()
  })

// Couple things to notice:
// - add/remove methods accept either a single value or an array
// - each add/remove is just queueing the operation,
//    no request is made until save() is called

// Deleting all permissions / resetting them would be done like so:
solid.getPermissions('https://www.example.com/file1')
  .then(function (permissionSet) {
    return permissionSet
      .clear()
      .save()
  })

// For existing permissions, clients can iterate through authorizations
// each authorization belongs to only one user or group (or everyone),
// and has multiple permission modes
solid.getPermissions('https://www.example.com/file1')
  .then(function (permissionSet) {
    permissionSet.forEach(function (authz) {    // loop through every authorization in an .acl
      if (authz.everyone()) { /* this is agentClass foaf:Agent */ }
      if (authz.group) { /* this is a group / agentClass */ }
      if (authz.agent) { /* this is for an individual agent */ }
      authz.iri  // -> webId of an agent or a group/agentClass
      // More than one of the following can be true:
      authz.allowsRead()  // -> true if authorization contains acl:Read mode
      authz.allowsWrite()  // -> true if acl:Write
      authz.allowsAppend()
      authz.allowsControl()
    })
  })
```

Note that there's no API for `defaultForNew` -- if the permission set is for a Container, then all authorizations are implicitly `defaultForNew`. If the permission set is for a regular non-container Resource, then it cannot have `defaultForNew`. Either way, the client doesn't need to know about it.

**Comment by ghanemabdo** [comment]
1) Currently, after the last refactoring, the permissions are keyed by authorization hash.
2) I totally agree, your new design to have cascading method calls for adding/removing auths is neat. I will have it on top of my list this week.

**Comment by dmitrizagidulin** [comment]
> 1) Currently, after the last refactoring, the permissions are keyed by authorization hash.

You're right. I look at it as essentially keyed by user, though. It uses a hash instead of the full agent's webId because it's shorter. And the `accessTo` resourceUrl is concatenated to the webId on the off chance that an `.acl` file will contain permissions for some other file (that's not its primary target). And so, in that one case, it makes sense to have multiple entries for one user. So the resourceUrl concatenation is to provide uniqueness.

**Comment by dmitrizagidulin** [comment]
Ok, merged the refactoring into this main `permissions` branch.
Heads up - I added another dev dependency, so be sure to run `npm install` when you fetch this branch.

**Comment by dmitrizagidulin** [comment]
Squashed commits, rebased on latest `master`.

**Comment by dmitrizagidulin** [comment]
Initial functionality implemented, README updated. Merging into master.
Thank you @ghanemabdo for all your help and hard work on this!

**Pull request merged**